src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,254 @@
+  rhs' <- analyseLocalSharing rhs
+  return (IR.App srcSpan lhs' rhs' typeScheme)
+
+-- | Checks an expression if it contains variables that occur
+--   multiple times on the same right-hand side.
+--   If that is the case, a @let@-expression is introduced that binds the
+--   variables to fresh ones and replaces the occurrences with the newly
+--   introduced variable.
+analyseSharingExpr :: IR.Expr -> Converter IR.Expr
+analyseSharingExpr expr = do
+  let varList
+        = (map fst . filter ((> 1) . snd) . Map.toList . countVarNamesExcept [])
+        expr
+  buildLet expr varList
+
+-- | Builds a @let@-expression from the given expression and variable names.
+--
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet :: IR.Expr -> [IR.VarName] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
+  let srcSpan = IR.exprSrcSpan expr
+  (binds, substs) <- buildBinds srcSpan vars
+  return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+          (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into @let@-bindings where
+--   the variable pattern is a fresh variable and the right-hand side is
+--   a variable that occurred multiple times on right hand sides.
+--
+--   Also computes substitutions mapping given variables to fresh variables.
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [IR.VarName] -> Converter ([IR.Bind], [Subst IR.Expr])
+buildBinds srcSpan = mapAndUnzipM buildBind
+ where
+  buildBind :: IR.VarName -> Converter (IR.Bind, Subst IR.Expr)
+  buildBind varName = do
+    varName' <- freshHaskellQName varName
+    let subst            = singleSubst' varName
+          (\s -> IR.Var s varName' Nothing)
+        -- This type scheme has wrong source span information.
+        rhs              = IR.Var srcSpan varName Nothing
+        Just varPatIdent = IR.identFromQName varName'
+        varPat           = IR.VarPat srcSpan varPatIdent Nothing False
+        bind             = IR.Bind srcSpan varPat rhs
+    return (bind, subst)
+
+-- | Counts all variable names on right-hand sides of expression.
+--   Shadowed variables and variables from the list are not counted.
+--   Variables introduced on the left side of a @case@-alternative and @let@
+--   expressions are not counted.
+countVarNamesExcept :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesExcept = countVarNamesWith (++) elem
+
+-- | Counts all variable names on right-hand sides of expression that are also
+--   contained by the given list.
+--   Shadowed variables and variables from the list are not counted.
+--   Variables introduced on the left side of a @case@-alternative and @let@
+--   expressions are not counted.
+countVarNamesOnly :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesOnly = countVarNamesWith const notElem
```

Comment:
The documentation states that

> Shadowed variables and variables from the list are not counted

but actually only variables from the list are counted. Shadowed variables are counted 
 as well since `const` ignores bound variables.